### PR TITLE
Refactor the test package-server

### PR DESCRIPTION
**Pull request by sigurdm**
Most of the action is in test/package_server.dart (and a bit in test/test_pub.dart).

### CODE REVIEW

File: test/add/common/version_resolution_test.dart [hunk 1]
```diff
@@ -41,18 +39,16 @@ void main() {
       () async {
     /// The server used to only have the foo v3.2.1 as the latest,
     /// so pub get will create a pubspec.lock to foo 3.2.1
-    await servePackages((builder) {
-      builder.serve('foo', '3.2.1');
-      builder.serve('bar', '1.0.0', deps: {'foo': '^3.2.1'});
-    });
+    final builder = await servePackages();
```
Comment (jonasfj) [inline test/add/common/version_resolution_test.dart:42]: Calling it `builder` is that a good name...

**Review by jonasfj — APPROVED** [review]

File: test/test_pub.dart [hunk 2]
```diff
@@ -956,3 +944,31 @@ Future<void> runPubIntoBuffer(
...
+
+/// The current global [PackageServer].
+PackageServer get globalServer => _globalServer!;
+PackageServer? _globalServer;
+
+/// Creates an HTTP server that replicates the structure of pub.dartlang.org and
+/// makes it the current [globalServer].
+///
+/// Calls [callback] with a [PackageServerBuilder] that's used to specify
+/// which packages to serve.
```
Comment (szakarias) [inline test/test_pub.dart:956]: Is this comment still up to date?

File: test/upgrade/report/does_not_show_newer_versions_for_locked_packages_test.dart [hunk 3]
```diff
@@ -11,14 +11,14 @@ void main() {
...
-      builder.serve('not_upgraded', '1.0.0');
-      builder.serve('not_upgraded', '2.0.0');
-      builder.serve('not_upgraded', '3.0.0-dev');
-      builder.serve('upgraded', '1.0.0');
-      builder.serve('upgraded', '2.0.0');
-      builder.serve('upgraded', '3.0.0-dev');
-    });
+    await servePackages()
+      ..serve('not_upgraded', '1.0.0')
+      ..serve('not_upgraded', '1.0.0')
```
Comment (szakarias) [inline test/upgrade/report/does_not_show_newer_versions_for_locked_packages_test.dart:16]: Delete?

File: test/upgrade/report/leading_character_shows_change_test.dart [hunk 4]
```diff
@@ -9,17 +9,17 @@ import '../../test_pub.dart';
...
-      builder.serve('overridden', '1.0.0');
-      builder.serve('removed', '1.0.0');
-      builder.serve('source_changed', '1.0.0');
-      builder.serve('upgraded', '1.0.0');
-      builder.serve('upgraded', '2.0.0');
-      builder.serve('unchanged', '1.0.0');
-    });
+    await servePackages()
+      ..serve('added', '1.0.0')
+      ..serve('added', '1.0.0')
```
Comment (szakarias) [inline test/upgrade/report/leading_character_shows_change_test.dart:14]: Delete?

File: test/upgrade/report/shows_newer_available_versions_test.dart [hunk 5]
```diff
@@ -9,28 +9,28 @@ import '../../test_pub.dart';
...
-      builder.serve('one_newer_unstable', '1.0.0');
-      builder.serve('one_newer_unstable', '1.0.1-unstable.1');
-      builder.serve('one_newer_unstable2', '1.0.1-unstable.1');
-      builder.serve('one_newer_unstable2', '1.0.1-unstable.2');
-      builder.serve('one_newer_stable', '1.0.0');
-      builder.serve('one_newer_stable', '1.0.1');
-    });
+    await servePackages()
+      ..serve('multiple_newer', '1.0.0')
+      ..serve('multiple_newer', '1.0.0')
```
Comment (szakarias) [inline test/upgrade/report/shows_newer_available_versions_test.dart:14]: Delete?

File: test/upgrade/report/shows_newer_available_versions_test.dart [hunk 6]
```diff
@@ -9,28 +9,28 @@ import '../../test_pub.dart';
...
+      ..serve('multiple_newer', '1.0.2-unstable.1')
+      ..serve('multiple_newer_stable', '1.0.0')
+      ..serve('multiple_newer_stable', '1.0.1')
+      ..serve('multiple_newer_stable', '1.0.2')
+      ..serve('multiple_newer_unstable', '1.0.0')
+      ..serve('multiple_newer_unstable', '1.0.1-unstable.1')
+      ..serve('multiple_newer_unstable', '1.0.1-unstable.2')
+      ..serve('multiple_newer_unstable2', '1.0.1-unstable.1')
+      ..serve('multiple_newer_unstable2', '1.0.1-unstable.2')
+      ..serve('multiple_newer_unstable2', '1.0.1-unstable.2')
```
Comment (szakarias) [inline test/upgrade/report/shows_newer_available_versions_test.dart:26]: This too?

**Review by szakarias — APPROVED** [review]
Very cool!

File: test/test_pub.dart (hunk 2 above)
Comment (sigurdm) [reply test/test_pub.dart]: Nope!

Thanks - fixed.

**Pull request merged**